### ISSUE TITLE
[BUG/Help] 如何修改modeling_chatglm.py 让self.model.stream_chat(self.tokenizer, prompt, history=history) 支持多线程

### ISSUE THREAD

**Issue by forever1dream** [THUDM/ChatGLM-6B#1160]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Current Behavior

自己用tornado 的websockt写了一个可以支持多用户同时访问的demo 但是发现一旦两个用户同时访问，结果就会很乱，和问题无关，单用户不会出现该情况，猜测应该是model.stram_chat本身不支持多线程，我应该如何修改
以下是websocket 的代码

```
import asyncio
import datetime
import json
import time

import tornado.websocket
import tornado.websocket
from tornado.options import define, parse_command_line
from tornado.platform.asyncio import AnyThreadEventLoopPolicy
from transformers import AutoTokenizer, AutoModel

define("port", default=8888, help="run on given port", type=int)

# 定义一个空的字典
clients = dict()


def build_prompt(history):
    prompt = "欢迎使用 ChatGLM-6B 模型，输入内容即可进行对话，clear 清空对话历史，stop 终止程序"
    q = ""
    r = ""
    for query, response in history:
        prompt += f"\n\nuser-input：{query}"
        prompt += f"\n\nChat-GLM-Answer：{response}"
        q = query
        r = response
    return prompt, q, r


class MyWebSocketHandler(tornado.websocket.WebSocketHandler):

    def initialize(self, my_object, my_tokenizer):
        self.model = my_object
        self.tokenizer = my_tokenizer
        self.executor = tornado.concurrent.futures.ThreadPoolExecutor(4)

    # 有新的连接时open()函数将会被调用,将客户端的连接统一放到clients
    def open(self, *args, **kwargs):
        self.id = self.get_argument("id")
        clients[self.id] = {"id": self.id, "object": self}
        print(clients)
        print("建立连接...")

    # @run_on_executor
    def predict(self, prompt, history):
        count = 0
        for response, history in self.model.stream_chat(self.tokenizer, prompt, history=history):
            count += 1
            print(count)
            if count % 4 == 0:
                p, q, r = build_prompt(history)
                self.write_message(r)
                print(r)

        p, q, r = build_prompt(history)
        self.write_message(r)

    # 客户收到消息将被调用
    # @gen.coroutine
    async def on_message(self, message):
        json_post = json.loads(message)
        prompt = json_post["query"]
        history = json_post['history']
        print(str(prompt))
        print(str(history))
        # yield self.predict(prompt=prompt, history=history)
        new_loop = asyncio.new_event_loop()
        asyncio.set_event_loop(new_loop)
        loop = asyncio.get_event_loop()
        await loop.run_in_executor(self.executor, self.predict, prompt, history)

    # 关闭连接时被调用
    def on_close(self):
        if self.id in clients:
            del clients[self.id]
        print("client %s is closed" % self.id)

    def check_origin(self, origin):
        return True


def send_time():
    while True:
        for key in clients.keys():
            msg = str(datetime.datetime.now())
            clients[key]["object"].write_message(msg)
            print("write to client %s:%s" % (key, msg))
        time.sleep(1)


if __name__ == "__main__":
    asyncio.set_event_loop_policy(AnyThreadEventLoopPolicy())
    # threading.Thread(target=send_time).start()

    tokenizer = AutoTokenizer.from_pretrained(".\\THUDM\\chatglm-6b-int4", trust_remote_code=True)
    model = AutoModel.from_pretrained(".\\THUDM\\chatglm-6b-int4", trust_remote_code=True).float()
    model.eval()
    app = tornado.web.Application([
        (r"/websocket", MyWebSocketHandler, {"my_object": model, "my_tokenizer": tokenizer})])

    parse_command_line()
    app.listen(80)
    tornado.ioloop.IOLoop.instance().start()

```

### Steps To Reproduce

多个用户同时访问

### Environment

```markdown
- OS:windows
- Python:3.7
- Transformers:4.27.1
- PyTorch:
- CUDA Support (`python -c "import torch; print(torch.cuda.is_available())"`) :
```

### Anything else?

no

**Comment by forever1dream** [comment]
> 多个线程同时访问共享数据，导致数据的不一致性。解决这个问题需要允许线程独占地访问共享数据，这就是线程同步。可以使用同步锁来解决这个问题。ChatGLM

我在stream_generate的while True中 forward前后加了threading.Lock().acquire()  和 threading.Lock().release() 依然是混乱的呢，能麻烦大佬说的再细点吗

**Comment by forever1dream** [comment]
解决了，多谢@rjk-git 可以参考： https://github.com/THUDM/ChatGLM-6B/issues/377
stream_generate方法的while True中forward的前面加锁（threadLock.acquire()）forward后解锁（threadLock.release()）保证只有一个线程跑forward，避免数据竞争。
具体修改如下：
在stream_generate函数中的 while true中
       ```
 while True:
            lock.acquire()
            model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
            # forward pass to get next token
            outputs = self(
                **model_inputs,
                return_dict=True,
                output_attentions=False,
                output_hidden_states=False,
            )
            lock.release()
            next_token_logits = outputs.logits[:, -1, :]
```

**Comment by rjk-git** [comment]
> 解决了，多谢@rjk-git 可以参考： #377 stream_generate方法的while True中forward的前面加锁（threadLock.acquire()）forward后解锁（threadLock.release()）保证只有一个线程跑forward，避免数据竞争。 具体修改如下： 在stream_generate函数中的 while true中 ``` while True: lock.acquire() model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs) # forward pass to get next token outputs = self( **model_inputs, return_dict=True, output_attentions=False, output_hidden_states=False, ) lock.release() next_token_logits = outputs.logits[:, -1, :]

不客气，熊熊～

**Comment by Lukangkang123** [comment]
> 多个线程同时访问共享数据，导致数据的不一致性。解决这个问题需要允许线程独占地访问共享数据，这就是线程同步。可以使用同步锁来解决这个问题。ChatGLM

独占地访问共享数据，推理时间不是成倍地增加了吗
